### PR TITLE
[xerces-c] Update to 3.2.2

### PR DESCRIPTION
**Pull request by rleigh-codelibre**
Hi,

My first vcpkg contribution, so there may well be some stylistic or other problems to correct.

- Update Xerces-C++ to 3.2.2.  This version provides a native CMake build, so use that to replace the (now removed) project/solution files
- Update Xalan to build with 3.2.x (primarily the C++11 `char16` patch)
- Xalan-C isn't being actively maintained, and there are a number of outstanding patches required to prevent segfaults and permit building on contemporary systems; I have included a few patches from JIRA here which are required.  Some are to permit building on Linux and MacOS X with followup changes to use the configure script when not building on Windows

Note that with regard to intellectual property rights, some of the patches for xalan were submitted upstream to the Xalan-C project and are under the terms of the Apache licence.  Most are my work, but some were created by Linux distribution vendors and submitted upstream:

- https://issues.apache.org/jira/browse/XALANC-747 (committed into subversion, but unreleased)
- https://issues.apache.org/jira/browse/XALANC-751 (committed into subversion, but unreleased)

i.e. they were already covered by the Apache CLA and passed upstream review.  The other patches are my own and are trivial one-liners, which are already in use by MacOS X homebrew, FreeBSD ports and some Linux distribution vendors, and have been tested and in use for some time.

The only other reverse dependency is `msix`.  This already embeds xerces-c 3.2.0 as vendored source, so it should work with 3.2.2 without trouble.  However, it failed to build with VS2015 with or without this change, so it seems like it might just be broken, independently of this PR?

Regards,
Roger

### CODE REVIEW

File: ports/xerces-c/CONTROL [hunk 1]
```diff
@@ -1,3 +1,3 @@
 Source: xerces-c
-Version: 3.2.2-1
+Version: 3.2.2-2
```
Comment (rleigh-codelibre) [inline ports/xerces-c/CONTROL]: Did the build number need incrementing twice in this update (both here, and also for xalan-c from `-1` to `-3`)?

**Comment by rleigh-codelibre** [comment]
@ras0219-msft Is there anything additional work you would like from me for this PR, or is it acceptable with the additional changes you made?

**Comment by rleigh-codelibre** [comment]
This PR should also allow building with `/std:c++17` due to the removal of `mem_fun`.

Are there any blocking issues for this PR?

**Comment by ras0219-msft** [comment]
Sorry for the long delay here, still trying to chase down a few more loose ends.

I don't expect any further changes needed on your side though; thank you for the PR!

**Comment by rleigh-codelibre** [comment]
Out of interest, what were the problems you saw on x64? Anything I should investigate upstream?

**Comment by ras0219-msft** [comment]
No, just a bug in our cmake testing helper -- it was always running under 32-bit mode, which means that package configs that insist on `CMAKE_SIZEOF_VOID_P` == `8` would fail :)

**Comment by ras0219-msft** [comment]
Thanks again for the PR and sorry once more for the long PR delay!

**Comment by rleigh-codelibre** [comment]
@ras0219-msft There's a problem with one of the changes here.  The `'LocalMsgIndex.hpp` header is missing.

```
FAILED: CMakeFiles/cmTC_c89f7.dir/src.cxx.obj 
C:\PROGRA~2\MICROS~1.0\VC\bin\amd64\cl.exe  /nologo /TP  -ID:\vcpkg-vs2015\installed\x64-windows\include /DWIN32 /D_WINDOWS /W3 /GR /EHsc /bigobj /W3 -DXALAN_LINK /MDd /Zi /Ob0 /Od /RTC1   -std:c++14 /showIncludes /FoCMakeFiles\cmTC_c89f7.dir\src.cxx.obj /FdCMakeFiles\cmTC_c89f7.dir\ /FS -c src.cxx
D:\vcpkg-vs2015\installed\x64-windows\include\xalanc/PlatformSupport/ProblemListenerBase.hpp(28): fatal error C1083: Cannot open include file: 'LocalMsgIndex.hpp': No such file or directory
ninja: build stopped: subcommand failed.

Return value: 1
Source file was:
#include <xercesc/util/PlatformUtils.hpp>
#include <xalanc/XalanTransformer/XalanTransformer.hpp>

int main()
{
  xercesc::XMLPlatformUtils::Initialize();
  xalanc::XalanTransformer::initialize();
  xalanc::XalanTransformer::terminate();
  xercesc::XMLPlatformUtils::Terminate();
}
```

I see it (and another header) in:

```
buildtrees\xalan-c\src\xalan-c-1.11\c\Build\Win64\VC10\Release\Nls\Include
buildtrees\xalan-c\src\xalan-c-1.11\c\Build\Win64\VC10\Debug\Nls\Include
```

```
D:\vcpkg>dir buildtrees\xalan-c\src\xalan-c-1.11\c\Build\Win32\VC10\Release\Nls\Include
 Volume in drive D is build
 Volume Serial Number is 5EDA-E3E0

 Directory of D:\vcpkg\buildtrees\xalan-c\src\xalan-c-1.11\c\Build\Win32\VC10\Release\Nls\Include

20/09/2018  15:30    <DIR>          .
20/09/2018  15:30    <DIR>          ..
20/09/2018  15:30            80,677 LocalMsgData.hpp
20/09/2018  15:30             9,285 LocalMsgIndex.hpp
               2 File(s)         89,962 bytes
               2 Dir(s)   7,123,279,872 bytes free

D:\vcpkg>dir buildtrees\xalan-c\src\xalan-c-1.11\c\Build\Win32\VC10\Debug\Nls\Include
 Volume in drive D is build
 Volume Serial Number is 5EDA-E3E0

 Directory of D:\vcpkg\buildtrees\xalan-c\src\xalan-c-1.11\c\Build\Win32\VC10\Debug\Nls\Include

20/09/2018  15:30    <DIR>          .
20/09/2018  15:30    <DIR>          ..
20/09/2018  15:30            80,677 LocalMsgData.hpp
20/09/2018  15:30             9,285 LocalMsgIndex.hpp
               2 File(s)         89,962 bytes
               2 Dir(s)   7,123,279,872 bytes free
```

I'm not sure what the purpose of the glob is here:

```
file(GLOB LOCALMSGINDEX ${CURRENT_BUILDTREES_DIR}/${VCPKG_TARGET_TRIPLET}-rel/*/c/Build/*/VC10/Release/Nls/Include/LocalMsgIndex.hpp)
```

It's not using the correct directory, and it's missing the other header (LocalMsgData.hpp).

**Comment by rleigh-codelibre** [comment]
Some investigation shows it's due to `${VCPKG_TARGET_TRIPLET}` being undefined.

**Comment by rleigh-codelibre** [comment]
After correcting it to `${TARGET_TRIPLET}`, this is still not working.  `vcpkg_install_msbuild` has deleted the source tree by this point, so there's nothing for the glob to find.

**Comment by rleigh-codelibre** [comment]
These issues should be fixed by https://github.com/Microsoft/vcpkg/pull/4565

This [build log](https://gitlab.com/rleigh/ome-common-cpp/-/jobs/112754200) is an example of the change being tested; note the `XALAN_LINK` line which compiles and runs the test program which was failing above.

**Pull request merged**